app/src/main/java/chat/rocket/android/chatroom/ui/ChatRoomAdapter.kt

```diff
@@ -3,41 +3,40 @@ package chat.rocket.android.chatroom.ui
 import android.support.v7.widget.RecyclerView
 import android.view.View
 import android.view.ViewGroup
-import android.widget.ImageView
 import android.widget.TextView
 import chat.rocket.android.R
 import chat.rocket.android.chatroom.viewmodel.AttachmentType
 import chat.rocket.android.chatroom.viewmodel.MessageViewModel
 import chat.rocket.android.player.PlayerActivity
 import chat.rocket.android.util.inflate
 import chat.rocket.android.util.setVisible
-import chat.rocket.common.util.ifNull
 import com.facebook.drawee.view.SimpleDraweeView
 import com.stfalcon.frescoimageviewer.ImageViewer
 import kotlinx.android.synthetic.main.avatar.view.*
 import kotlinx.android.synthetic.main.item_message.view.*
 import kotlinx.android.synthetic.main.message_attachment.view.*
 
-class ChatRoomAdapter(private val serverUrl: String) : RecyclerView.Adapter<ChatRoomAdapter.ViewHolder>() {
+class ChatRoomAdapter: RecyclerView.Adapter<ChatRoomAdapter.ViewHolder>() {
+    private val dataSet = ArrayList<MessageViewModel>()
 
     init {
         setHasStableIds(true)
     }
 
-    val dataSet = ArrayList<MessageViewModel>()
-
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
-            ViewHolder(parent.inflate(R.layout.item_message), serverUrl)
+            ViewHolder(parent.inflate(R.layout.item_message))
 
-    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(dataSet[position])
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
+        holder.bind(dataSet[position], dataSet.getOrNull(position + 1))
 
-    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: MutableList<Any>?) {
-        onBindViewHolder(holder, position)
-    }
+    override fun getItemCount(): Int =
+        dataSet.size
 
-    override fun getItemCount(): Int = dataSet.size
+    override fun getItemViewType(position: Int): Int =
```

Comment:
That way you will never recycle a View on the RecyclerView.

https://developer.android.com/reference/android/support/v7/widget/RecyclerView.Adapter.html#getItemViewType(int)

Reply:
? both `getItemCount()` and `getItemViewType()` is here. 
What are you referring to?

Reply:
? both `getItemCount()` and `getItemViewType()` is here. 
What are you referring to?

Reply:
```kotlin
    override fun getItemViewType(position: Int): Int =
        position
```
You are using the position for the View Type, each row will have a different view type, preventing the RecyclerView to recycler any view.

Reply:
And we are actually not using different view types here.